gcloud/pubsub/topic.py

```diff
@@ -0,0 +1,149 @@
+    def exists(self):
+        """API call:  test for the existence of the topic via a GET request
+
+        See
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/topics/get
+        """
+        try:
+            self.connection.api_request(method='GET', path=self.path)
+        except NotFound:
+            return False
+        else:
+            return True
+
+    def publish(self, message, **attrs):
+        """API call:  publish a message to a topic via a POST request
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/topics/publish
+
+        :type message: bytes
+        :param message: the message payload
+
+        :type attrs: dict (string -> string)
+        :message attrs: key-value pairs to send as message attributes
+
+        :rtype: str
+        :returns: message ID assigned by the server to the published message
+        """
+        message_data = {'data': base64.b64encode(message), 'attributes': attrs}
+        data = {'messages': [message_data]}
+        response = self.connection.api_request(method='POST',
+                                               path='%s:publish' % self.path,
+                                               data=data)
+        return response['messageIds'][0]
+
+    def batch(self):
+        """Return a batch to use as a context manager.
+
+        :rtype: :class:_Batch
+        """
+        return _Batch(self)
+
+    def delete(self):
+        """API call:  delete the topic via a DELETE request
+
+        See:
+        https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/topics/delete
+        """
+        self.connection.api_request(method='DELETE', path=self.path)
+
+
+class _Batch(object):
+    """Context manager:  collect messages to publish via a single API call.
+
+    Helper returned by :meth:Topic.batch
+    """
+    def __init__(self, topic):
+        self.topic = topic
+        self.messages = []
+        self.message_ids = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            self.commit()
+
+    def __iter__(self):
+        return iter(self.message_ids)
+
+    def publish(self, message, **attrs):
+        """Emulate publishing a message, but save it.
+
+        :type message: bytes
+        :param message: the message payload
+
+        :type attrs: dict (string -> string)
+        :message attrs: key-value pairs to send as message attributes
+        """
+        self.messages.append(
+            {'data': base64.b64encode(message), 'attributes': attrs})
+
+    def commit(self):
+        """Send saved messages as a single API call."""
+        conn = self.topic.connection
+        response = conn.api_request(method='POST',
+                                    path='%s:publish' % self.topic.path,
+                                    data={'messages': self.messages})
+        self.message_ids.extend(response['messageIds'])
```

Comment:
Shouldn't `self.messages` be cleaned up after being committed successfully? Or set a bool on the `_Batch` to tombstone it after calling `commit`.